zeus/run/dataloader.py

```diff
@@ -690,45 +909,57 @@ def __next__(self):
+                    # record the tput/power of the optimal PL. From the following epochs where we
+                    # don't profile anything, we directly use these values to compute the time and
+                    # energy consumed.
+                    if self._should_profile:
+                        self.eval_tput_result[self.current_gpu_pl] = (
+                            self.num_samples / time_consumption
+                        )
+                        self.eval_power_result[self.current_gpu_pl] = (
+                            sum_energy_consumption / time_consumption
+                        )
+                        # The optimal PL being known means that all power limits have been explored.
+                        # Let us end profiling by writing profile information to `power_json`.
+                        if self.optimal_pl != 0:
+                            self._save_power_results()
+                self._log(
+                    f"{self.split} epoch {self.epoch_num} done: "
+                    f"time={time_consumption:.2f} energy={energy_consumption:.2f}"
                 )
-                self.eval_epoch_time.append(time_consumption)
-                self.eval_epoch_energy.append(energy_consumption)
-                # For the eval dataloader, we want to record the throughput and power
-                # for the current power limit. Since the train dataloader sets the power limit
-                # to the optimal power limit right after profiling is done, this will naturally
-                # record the tput/power of the optimal PL. From the following epochs where we
-                # don't profile anything, we directly use these values to compute the time and
-                # energy consumed.
-                if self._should_profile:
-                    self.eval_tput_result[self.current_gpu_pl] = (
-                        self.num_samples / time_consumption
-                    )
-                    self.eval_power_result[self.current_gpu_pl] = (
-                        energy_consumption / time_consumption
-                    )
-                    # The optimal PL being known means that all power limits have been explored.
-                    # Let us end profiling by writing profile information to `power_json`.
-                    if self.optimal_pl != 0:
-                        self._save_power_results()
-            self._log(
-                f"{self.split} epoch {self.epoch_num} done: "
-                f"time={time_consumption:.2f} energy={energy_consumption:.2f}"
-            )
 
-            # Epoch ended in the middle of a profiling window. Reset.
-            if self._is_train and self.prof_state != NOT_PROFILING:
-                self._end_prof(cutoff=True)
+                # Epoch ended in the middle of a profiling window. Reset.
+                if self._is_train and self.prof_state != NOT_PROFILING:
+                    self._end_prof(cutoff=True)
```

Comment:
Not sure about the exact implication of this, but `self._end_prof` here is only called for the master process (rank 0). On the contrary, in lines 968-983, `self._start_warmup`, etc are called for all ranks, including `self._end_prof`. Just wanted to check whether there's nothing wrong with this difference.

Reply:
I re-think about it. I now think we have to let all the processes enter in this case, because even `cutoff==True`, the piece of code in `self._end_prof` still does an important job: set the state back to `NOT_PROFILING`. Otherwise, the master process will start from warmup but the others won't, which is bad. This is also because we don't have a set state back to `NOT_PROFILING` in `__init__`, I think.

Regarding to the effect of terminating all the processes once cutoff twice, I think it will be the same for the two.